src/extensions/sb-ncbr/partial-charges/property.ts

```diff
@@ -0,0 +1,193 @@
+                typeIdToResidueToCharge,
+                maxAbsoluteAtomCharges,
+                maxAbsoluteResidueCharges,
+                maxAbsoluteAtomChargeAll,
+            },
+        },
+    };
+}
+
+function getTypeIdToMethod(model: Model) {
+    const typeIdToMethod: ChargesData['typeIdToMethod'] = new Map();
+
+    const sourceData = model.sourceData as MmcifFormat;
+    const rowCount = sourceData.data.frame.categories.partial_atomic_charges_meta.rowCount;
+    const typeIds = sourceData.data.frame.categories.partial_atomic_charges_meta.getField('id')?.toIntArray();
+    const methods = sourceData.data.frame.categories.partial_atomic_charges_meta.getField('method')?.toStringArray();
+
+    if (!typeIds || !methods) {
+        return typeIdToMethod;
+    }
+
+    for (let i = 0; i < rowCount; ++i) {
+        const typeId = typeIds[i];
+        const method = methods[i];
+        typeIdToMethod.set(typeId, method);
+    }
+
+    return typeIdToMethod;
+}
+
+function getTypeIdToAtomIdToCharge(model: Model): ChargesData['typeIdToAtomIdToCharge'] {
+    const atomIdToCharge: ChargesData['typeIdToAtomIdToCharge'] = new Map();
+
+    const sourceData = model.sourceData as MmcifFormat;
+    const rowCount = sourceData.data.frame.categories.partial_atomic_charges.rowCount;
+    const typeIds = sourceData.data.frame.categories.partial_atomic_charges.getField('type_id')?.toIntArray();
+    const atomIds = sourceData.data.frame.categories.partial_atomic_charges.getField('atom_id')?.toIntArray();
+    const charges = sourceData.data.frame.categories.partial_atomic_charges.getField('charge')?.toFloatArray();
+
+    if (!typeIds || !atomIds || !charges) return atomIdToCharge;
+
+    for (let i = 0; i < rowCount; ++i) {
+        const typeId = typeIds[i];
+        const atomId = atomIds[i];
+        const charge = charges[i];
+        if (!atomIdToCharge.has(typeId)) atomIdToCharge.set(typeId, new Map());
+        atomIdToCharge.get(typeId)?.set(atomId, charge);
+    }
+
+    return atomIdToCharge;
+}
+
+function getTypeIdToResidueIdToCharge(model: Model, typeIdToAtomIdToCharge: ChargesData['typeIdToAtomIdToCharge']) {
+    const { offsets } = model.atomicHierarchy.residueAtomSegments;
+    const residueToCharge: ChargesData['typeIdToResidueToCharge'] = new Map();
+
+    typeIdToAtomIdToCharge.forEach((atomIdToCharge, typeId: number) => {
+        for (let residueId = 1; residueId < offsets.length; ++residueId) {
+            let charge = 0;
+            for (let atomId = offsets[residueId - 1] + 1; atomId <= offsets[residueId]; ++atomId) {
+                charge += atomIdToCharge?.get(atomId) || 0;
+            }
+            for (let atomId = offsets[residueId - 1] + 1; atomId <= offsets[residueId]; ++atomId) {
+                if (!residueToCharge.has(typeId)) residueToCharge.set(typeId, new Map());
+                residueToCharge.get(typeId)?.set(atomId, Number(charge.toFixed(4)));
+            }
+        }
```

Comment:
How did you arrive at this usage of the offsets? This is not how they work. Also, I assume your `partial_atomic_charges.atom_id` should map to `atom_site.label_atom_id` and in that case, you mapping doesn't always have to be correct. I assume `offsets[residueId - 1] + 1` was a hack to get the 1-based indexing you see in most mmCIF files.

I am writing this from the top of my head, but it should be something like this:

```ts
const { offsets, count } = model.atomicHierarchy.residueAtomSegments;
const atomId = StructureProperties.atom.label_atom_id;
const label_atom_id = model.atomicHierarchy.atoms.label_atom_id;
for (let rI = 0; rI < count; rI++) {
  for (let aI = offsets[rI], _aI = offsets[rI+1]; aI < _aI; aI++) {
    const atom_id = label_atom_id.value(aI);
     // ...
  }
}
```

Reply:
I'm getting the residue charge by summing all charges of atoms in the residue. The reason why I'm then writing the resulting residue charge to each atomId in the residue is because I don't know which atomId Molstar will use when coloring residues in the ColorTheme.

The `partial_atomic_charges.atom_id` item should map to `atom_site.id`, however I wasn't able to find it in model.atomicHierarchy.

Reply:
Right, I made a mistake with the field name. You can still access the id column via the raw "atom_site" and use `sourceIndex` from the hierarchy to get the correct index into that category.

In general your current approach wont work because mol* also sorts the inputs (hence the existence of the `sourceIndex`).